overlord/snapstate/target.go

```diff
@@ -677,15 +704,53 @@ func (p *pathInstallGoal) toInstall(ctx context.Context, st *state.State, opts O
 		return nil, err
 	}
 
+	comps, err := componentSetupsFromPaths(info, p.components)
+	if err != nil {
+		return nil, err
+	}
+
 	inst := target{
 		setup: SnapSetup{
 			SnapPath:  p.path,
 			Channel:   channel,
 			CohortKey: p.revOpts.CohortKey,
 		},
-		info:   info,
-		snapst: snapst,
+		info:       info,
+		snapst:     snapst,
+		components: comps,
 	}
 
 	return []target{inst}, nil
 }
+
+func componentSetupsFromPaths(snapInfo *snap.Info, components map[*snap.ComponentSideInfo]string) ([]ComponentSetup, error) {
+	setups := make([]ComponentSetup, 0, len(components))
+	for csi, path := range components {
+		compInfo, err := validatedComponentInfo(path, snapInfo, csi)
+		if err != nil {
+			return nil, err
+		}
+
+		setups = append(setups, ComponentSetup{
+			CompPath:     path,
+			CompSideInfo: csi,
+			CompType:     compInfo.Type,
+		})
+	}
+
+	return setups, nil
+}
+
+func validatedComponentInfo(path string, si *snap.Info, csi *snap.ComponentSideInfo) (*snap.ComponentInfo, error) {
+	if err := csi.Component.Validate(); err != nil {
+		return nil, err
```

Comment:
this path doesn't seem to fail in tests?  also where do we check that the snap name and snap name in the component match?  is something we do later in the tasks? maybe we should do that early as well?

Reply:
That error is covered by TestInstallComponentsFromPathInvalidComponentName.

Reply:
We check that the snap name and the snap name in the component match inside of [`snap.ReadComponentInfoFromContainer`](https://github.com/snapcore/snapd/blob/master/snap/component.go?plain=1#L228), which is called by `OpenComponentFile`.